File: Java/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     *@param preorder : A list of integers that preorder traversal of a tree
     *@param inorder : A list of integers that inorder traversal of a tree
     *@return : Root of a tree
     */
    public TreeNode helper(int[] preorder, int[] inorder, int l1, int r1, int l2, int r2) {
        if (l1 > r1) return null;
        if (l1 == r1) {
            TreeNode root = new TreeNode(preorder[l1]);
            return root;
        }
        int cur = 0;
        for (int i = l2; i <= r2; i ++) {
            if (inorder[i] == preorder[l1]) {
                cur = i;
                break;
            }
        }
        TreeNode root = new TreeNode(inorder[cur]);
        root.left = helper(preorder, inorder, l1 + 1, l1 + cur - l2, l2, cur - 1);
        root.right = helper(preorder, inorder, l1 + cur - l2 + 1, r1, cur + 1, r2);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        // write your code here
        int len = preorder.length;
        return helper(preorder, inorder, 0, len - 1, 0, len - 1);
    }
}

File: Cpp/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.cpp
#include <vector>

/**
 * Definition of TreeNode:
 **/
class TreeNode {
public:
    int val;
    TreeNode *left, *right;
    TreeNode(int val) {
        this->val = val;
        this->left = this->right = NULL;
    }
};

class Solution {
    /**
     *@param inorder : A list of integers that inorder traversal of a tree
     *@param postorder : A list of integers that postorder traversal of a tree
     *@return : Root of a tree
     */
public:
    TreeNode *build(vector<int> &inorder, vector<int> &preorder, int pos1, int pos2, int pos3, int pos4) {
        if (pos1 > pos2) return NULL;
        if (pos1 == pos2) {
            TreeNode *root = new TreeNode(inorder[pos1]);
            return root;
        }
        int index;
        for (int i = pos2; i >= pos1; i --) {
            if (preorder[pos3] == inorder[i]) {
                index = i;
                break;
            }
        }
        TreeNode *root = new TreeNode(preorder[pos3]);
        root->left = build(inorder, preorder, pos1, index - 1, pos3 + 1, pos3 + index - pos1);
        root->right = build(inorder, preorder, index + 1, pos2, pos3 + index - pos1 + 1, pos4);
        return root;
    }
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        // write your code here
        if (inorder.size() == 0) return NULL;
        return build(inorder, preorder, 0, inorder.size() - 1, 0, preorder.size() - 1);
    }
};
